compiler/rustc_typeck/src/check_unused.rs

```diff
@@ -16,48 +16,37 @@ pub fn check_crate(tcx: TyCtxt<'_>) {
         used_trait_imports.extend(imports.iter());
     }
 
-    for id in tcx.hir().items() {
-        if matches!(tcx.def_kind(id.def_id), DefKind::Use) {
-            if tcx.visibility(id.def_id).is_public() {
-                continue;
-            }
-            let item = tcx.hir().item(id);
-            if item.span.is_dummy() {
-                continue;
-            }
-            if let hir::ItemKind::Use(path, _) = item.kind {
-                check_import(tcx, &mut used_trait_imports, item.item_id(), path.span);
-            }
+    for &id in tcx.maybe_unused_trait_imports(()) {
+        debug_assert!(tcx.def_kind(id) == DefKind::Use);
+        if tcx.visibility(id).is_public() {
+            continue;
         }
+        if used_trait_imports.contains(&id) {
+            continue;
+        }
+        let item = tcx.hir().expect_item(id);
+        // if item.span.is_dummy() {
+        //     continue;
+        // }
+        let hir::ItemKind::Use(path, _) = item.kind else { unreachable!() };
+        tcx.struct_span_lint_hir(
+            lint::builtin::UNUSED_IMPORTS,
+            tcx.local_def_id_to_hir_id(id),
```

Comment:
This is equal to `item.hir_id()`.

Reply:
Ahhh, thank you. This was why I was asking if I did anything silly with the IDs. I just grepped for type signatures that matched the conversion I wanted haha. I'll get the hang of it eventually.